### PR TITLE
[Bugfix] Fix prompt_embeds precision divergence with MTP speculative …

### PR DESCRIPTION
**Pull request by leihuang-sketch**
Hi from [novita.ai](https://novita.ai/) team 👋

When --enable-prompt-embeds and --speculative-config.method mtp are both enabled, logits from prompt_embeds input diverge from text input starting at the 2nd generated token. Two root causes:

1. Async scheduling scatter paths update input_ids.gpu but not is_token_ids.gpu, causing the target model to use stale embeddings instead of re-embedding from token ids for decode positions.

2. MTP drafter re-embeds from token ids instead of using the target model's actual inputs_embeds, producing inconsistent embeddings for prompt_embeds inputs.

Also adds a startup validation rejecting --enable-prompt-embeds with multimodal models, as they have incompatible embedding pipelines.

## Test Plan
test on [glm5-fp8](https://huggingface.co/zai-org/GLM-5-FP8), vllm serve

```
nohup vllm serve /path/to/GLM-5-FP8/ \
     --port 33333  \
     --tensor-parallel-size 8  \
     --runner generate   \
     --max-model-len 8192 \
     --enable-prompt-embeds \
     --speculative-config.method mtp \
     --speculative-config.num_speculative_tokens 1 \
     --tool-call-parser glm47 \
     --reasoning-parser glm45 \
     --enable-auto-tool-choice \
     --served-model-name glm5 > serve.log &
```

## Test Result
Test from [this example](https://github.com/vllm-project/vllm/blob/main/examples/online_serving/prompt_embed_inference_with_openai_client.py)
main branch：
```
------------------------------
11.
------------------------------
```

My branch has fixed the accuracy issue
```
------------------------------
1.  **Analyze the Request:** The user is asking for information about the capital of France.

2.  **Identify the Subject:** The capital of France is Paris.

3.  **Determine Key Information Categories:** To provide a comprehensive answer, I should cover several aspects of Paris
------------------------------
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

File: vllm/v1/spec_decode/eagle.py [hunk 1]
```diff
@@ -293,6 +296,13 @@ def _raise_if_multimodal(self):
                 "does not support multimodal models yet"
             )
 
+    def _raise_if_prompt_embeds(self):
+        if self.supports_prompt_embeds:
+            raise NotImplementedError(
```
Comment (benchislett) [inline vllm/v1/spec_decode/eagle.py:301]: Can you elaborate on what the blocker is for enablement here?

Comment (leihuang-sketch) [reply vllm/v1/spec_decode/eagle.py:301]: @benchislett 
When needs_extra_input_slots is true (i.e., EAGLE with a separate draft model or parallel drafting), the copy_and_expand_eagle_inputs_kernel triton kernel is responsible for reshaping the input tensors — it copies input_ids and positions into expanded buffers, inserting extra padding slots per request (filled with padding_token_id=0), and recomputes the slot mapping accordingly.

This kernel operates entirely on integer tensors (input_ids and positions). With prompt embeds enabled, the actual model input is a high-dimensional float inputs_embeds tensor rather than input_ids. The kernel has no logic to:

Pad the inputs_embeds tensor with appropriate dummy embeddings for the extra slots (you can't just fill with a scalar 0 like padding_token_id — you'd need a full embedding vector).
Shift/rotate the inputs_embeds tensor the same way it shifts input_ids.
Compute out_hidden_state_mapping for the embedding space.
For reference, the non-needs_extra_input_slots path (used by MTP) already handles prompt embeds correctly (lines 700-705 in eagle.py) — it shifts inputs_embeds in-place and uses embed_input_ids() to convert sampled next tokens into embeddings. But that logic is simple because it doesn't need to insert any padding slots.

To enable prompt embeds in the needs_extra_input_slots path, we would need to either extend the triton kernel to also handle the inputs_embeds tensor, or add a separate post-kernel step that pads and rearranges inputs_embeds to match the expanded layout. This is doable but non-trivial, so we guard it with NotImplementedError for now.

Comment (leihuang-sketch) [reply vllm/v1/spec_decode/eagle.py:301]: image

**Review by qthequartermasterman — COMMENTED** [review]
This is super cool. Does this mean that Eagle speculative decoding is fully compatible with prompt embeds after this PR? 

If so, could you please update the compatibility matrix in the docs with the appropriate details? https://docs.vllm.ai/en/latest/features/?h=features#feature-x-feature . 

Also are there any regression tests showing prompt embeds + eagle speculative decoding?

File: vllm/config/model.py [hunk 2]
```diff
@@ -628,6 +628,13 @@ def __post_init__(
                 "repo name or path using the --tokenizer argument."
             )
 
+        if self.enable_prompt_embeds and self.is_multimodal_model:
```
Comment (LuisRobaina) [inline vllm/config/model.py:631]: Hi, this is not the case anymore after v0.20.1. Now `prompt_embeds` can be used with multimodal models.

See for example: https://github.com/vllm-project/vllm/blob/main/tests/entrypoints/openai/chat_completion/test_chat_completion_with_mixed_image_embeds.py

**Comment by leihuang-sketch** [comment]
> This is super cool. Does this mean that Eagle speculative decoding is fully compatible with prompt embeds after this PR?
> 
> If so, could you please update the compatibility matrix in the docs with the appropriate details? https://docs.vllm.ai/en/latest/features/?h=features#feature-x-feature .
> 
> Also are there any regression tests showing prompt embeds + eagle speculative decoding?

@qthequartermasterman At present, embedding+eagle has not been tested, only embedding+mtp has been tested. Pull the latest code to supplement the testing of this scenario

**Comment by leihuang-sketch** [comment]
@qthequartermasterman @LuisRobaina close it, and new pr #42168
